paasta_tools/kubernetes/application/controller_wrappers.py

```diff
@@ -188,14 +203,173 @@ def deep_delete(self, kube_client: KubeClient) -> None:
                 )
             )
         self.delete_pod_disruption_budget(kube_client)
+        self.delete_horizontal_pod_autoscaler(kube_client)
+
+    def get_existing_app(self, kube_client: KubeClient):
+        return kube_client.deployments.read_namespaced_deployment(
+            name=self.item.metadata.name, namespace=self.item.metadata.namespace
+        )
 
     def create(self, kube_client: KubeClient) -> None:
         create_deployment(kube_client=kube_client, formatted_deployment=self.item)
         self.ensure_pod_disruption_budget(kube_client)
+        self.sync_horizontal_pod_autoscaler(kube_client)
 
     def update(self, kube_client: KubeClient) -> None:
+        # If autoscaling is enabled, do not update replicas.
+        # In all other cases, replica is set to max(instances, min_instances)
+        if not self.get_soa_config().get("instances"):
```

Comment:
I think this is a little confusing. How about if we override `get_instances` in `KubernetesDeploymentConfig` and maybe have an `initial_instances`? I don't 100% know how the HPA works but my guess is that you just set `replicas` to some initial value and then the HPA is used to get the real value?

Reply:
HPA updates the replicas parameter proportionally based on metrics. So we would not update replicas here. Ideally, we should just use patch_namespaced_deployment for update, but I'm not sure if that's a good idea here.  
And sure I can override get_instances in KubernetesDeploymentConfig, but KubernetesDeploymentConfig does not know if the formatted_application is used for update or create. So I either need to refactor the code to make it know, or send a request to k8s API to check if it exists or not. Refactoring the format_application is dirtier than what I do now; Sending requests to k8s wastes network resources.

Reply:
Use patch instead of replace is probably the ideal solution here. What do you think?

Reply:
I think you may still need to implement get_instances to reflect autoscaling because it's used in other places - for example, `check_kubernetes_services_replication.py`. But I don't have an opinion on whether you should use that or create a method to patch the deployment.